src/libsystemd-network/dhcp-option.c

```diff
@@ -93,32 +112,93 @@ static int option_append(uint8_t options[], size_t size, size_t *offset,
+                sd_dhcp_server *server = (sd_dhcp_server *) optval;
+                size_t current_offset = *offset + 2;
 
+                if (server->agent_circuit_id) {
+                        r = dhcp_option_append_tlv(options, size, &current_offset, SD_DHCP_RELAY_AGENT_CIRCUIT_ID,
+                                                   strlen(server->agent_circuit_id), server->agent_circuit_id);
+                        if (r < 0)
+                                return r;
+                }
+                if (server->agent_remote_id) {
+                        r = dhcp_option_append_tlv(options, size, &current_offset, SD_DHCP_RELAY_AGENT_REMOTE_ID,
+                                                   strlen(server->agent_remote_id), server->agent_remote_id);
+                        if (r < 0)
+                                return r;
+                }
+
+                options[*offset] = code;
+                options[*offset + 1] = current_offset - *offset - 2;
+                assert(current_offset - *offset - 2 <= 255);
+                *offset = current_offset;
                 break;
         }
         default:
-                if (*offset + 2 + optlen > size)
-                        return -ENOBUFS;
+                return dhcp_option_append_tlv(options, size, offset, code, optlen, optval);
+        }
+        return 0;
+}
 
-                options[*offset] = code;
-                options[*offset + 1] = optlen;
+static int option_length(uint8_t *options, size_t length, size_t offset) {
+        assert(options);
+        assert(offset < length);
 
-                memcpy_safe(&options[*offset + 2], optval, optlen);
-                *offset += 2 + optlen;
+        if (options[offset] == SD_DHCP_OPTION_PAD || options[offset] == SD_DHCP_OPTION_END)
+                return 1;
+        if (length < offset + 2)
+                return -ENOBUFS;
 
-                break;
+        //validating that buffer is long enough
+        if (length < offset + 2 + options[offset + 1])
+                return -ENOBUFS;
+
+        return options[offset + 1] + 2;
+}
+
+int dhcp_option_find_option(uint8_t *options, size_t length, uint8_t wanted_code) {
+        assert(options);
+        size_t offset = 0;
+        int r;
+
+        while (offset < length) {
+                if (wanted_code == options[offset])
+                        return offset;
+
+                r = option_length(options, length, offset);
+                if (r < 0)
+                        return r;
+                offset += r;
         }
+        return offset == length ? -ENOENT : -ENOBUFS;
+}
 
-        return 0;
+int dhcp_option_remove_option(uint8_t *options, size_t length, uint8_t option_code) {
+        int r;
+        int s;
+
+        assert(options);
+
+        r = dhcp_option_find_option(options, length, option_code);
+        if (r < 0)
+                return r;
+
+        s = option_length(options, length, r);
+        if (s < 0)
+                return s;
```

Comment:
`dhcp_option_find_option()` already calls `option_length()`. So, it is better that `dhcp_option_find_option()` also returns the length of found option.